In [65]:
# Importig Pandas
import pandas as pd

# Importig Matplotlib
import matplotlib.pyplot as plt

# Importing Numpy
import numpy as np


In [66]:
# Import the requests library.
import requests

# Import the API key.
from config import weather_api_key

In [67]:
# Build the basic URL for the OpenWeatherMap with your weather-
#_api_key added to the URL.
url = "http://api.openweathermap.org/data/2.5/weather?units=Imperial&APPID=" + weather_api_key

# Declares the url variable.
print(url)

# Declares the city_url variable.
city_url= url + "&q=" + "Boston"

# Prints the city_url variable.
print(city_url)

http://api.openweathermap.org/data/2.5/weather?units=Imperial&APPID=47ea430ffbc6a3d7f5e1667d3293f52a
http://api.openweathermap.org/data/2.5/weather?units=Imperial&APPID=47ea430ffbc6a3d7f5e1667d3293f52a&q=Boston


In [68]:
# Import the datetime module from the datetime library.
from datetime import datetime

In [69]:
# Creates an empty list to hold the  data.
city_data=[]

# Print the beginning of the logging.
print("Beginning Data Retrieval  ")
print("--------------------------")

# Create the counters.
# we have initialized the counters at 1 because we want the 
#first iteration of the logging for each recorded response
record_count=1

# Set to start at 1.
set_count=1

Beginning Data Retrieval  
--------------------------


In [70]:
# Loop through all the cities in our list.
for i in range(len(cities)):

    # Group cities in sets of 50 for logging purposes.
    if (i % 50 == 0 and i >= 50):
        set_count += 1
        record_count = 1
    # Create endpoint URL with each city.
    city_url = url + "&q=" + "i"

In [71]:
import numpy as np

np.random.seed(42)
lats = np.random.uniform(low=-90.000, high=90.000, size=1500)
lngs = np.random.uniform(low=-180.000, high=180.000, size=1500)
lat_lngs = list(zip(lats, lngs))
lat_lngs 

# Use the citipy module to determine city based on latitude and longitude.
from citipy import citipy

# Create a list for holding the cities.
cities = []
# Identify the nearest city for each latitude and longitude combination.
for lat_lng in lat_lngs:
    city = citipy.nearest_city(lat_lng[0], lat_lng[1]).city_name

    # If the city is unique, then we will add it to the cities list.
    if city not in cities:
        cities.append(city)
# Print the city count to confirm sufficient count.
len(cities)

597

In [72]:

print(url)

http://api.openweathermap.org/data/2.5/weather?units=Imperial&APPID=47ea430ffbc6a3d7f5e1667d3293f52a


In [73]:
city_data = []
# Loop through all the cities in the list.
for i, city in enumerate(cities):

    # Group cities in sets of 50 for logging purposes.
    if (i % 50 == 0 and i >= 50):
        set_count += 1
        record_count = 1
    # Create endpoint URL with each city.
    city_url = url + "&q=" + city

    
    # Run API request for each of the cities.
    try:
        # Parse the JSON and retrieve data.
        city_weather = requests.get(city_url).json()
        # Parse out the needed data.
        city_lat = city_weather["coord"]["lat"]
        city_lng = city_weather["coord"]["lon"]
        city_max_temp = city_weather["main"]["temp_max"]
        city_humidity = city_weather["main"]["humidity"]
        city_clouds = city_weather["clouds"]["all"]
        city_wind = city_weather["wind"]["speed"]
        city_country = city_weather["sys"]["country"]
        # Convert the date to ISO standard.
        city_date = datetime.utcfromtimestamp(city_weather["dt"]).strftime('%Y-%m-%d %H:%M:%S')
        # Append the city information into city_data list.
        city_data.append({"City": city.title(),
                          "Lat": city_lat,
                          "Lng": city_lng,
                          "Max Temp": city_max_temp,
                          "Humidity": city_humidity,
                          "Cloudiness": city_clouds,
                          "Wind Speed": city_wind,
                          "Country": city_country,
                          "Date": city_date})

        print('City found , max temp was ', city_max_temp)
# If an error is experienced, skip the city.
    except:
        print("City not found. Skipping...")
        pass

# Indicate that Data Loading is complete.
print("-----------------------------")
print("Data Retrieval Complete      ")
print("-----------------------------")
print(city_data)

City found , max temp was  56.66
City not found. Skipping...
City found , max temp was  37.4
City found , max temp was  46.4
City found , max temp was  33.8
City found , max temp was  49.62
City found , max temp was  56.61
City found , max temp was  32
City not found. Skipping...
City found , max temp was  44.55
City found , max temp was  39
City found , max temp was  -11.64
City found , max temp was  38.97
City found , max temp was  69.8
City found , max temp was  35.46
City not found. Skipping...
City found , max temp was  82.54
City found , max temp was  64.2
City found , max temp was  75.2
City found , max temp was  86
City found , max temp was  88.39
City found , max temp was  65.37
City found , max temp was  66.67
City found , max temp was  65.84
City found , max temp was  15.8
City found , max temp was  37.4
City found , max temp was  38.34
City not found. Skipping...
City found , max temp was  59
City not found. Skipping...
City found , max temp was  61
City found , max temp wa

City found , max temp was  65.73
City found , max temp was  28.4
City not found. Skipping...
City found , max temp was  26.46
City found , max temp was  46.99
City not found. Skipping...
City found , max temp was  84.2
City found , max temp was  66.72
City not found. Skipping...
City found , max temp was  83.71
City found , max temp was  77
City found , max temp was  35.46
City found , max temp was  78.8
City found , max temp was  12.2
City found , max temp was  67.91
City found , max temp was  24.8
City found , max temp was  55.09
City found , max temp was  34.25
City found , max temp was  73.49
City not found. Skipping...
City found , max temp was  61
City found , max temp was  50
City found , max temp was  70.45
City found , max temp was  64.24
City found , max temp was  53.01
City found , max temp was  81.28
City found , max temp was  37.99
City found , max temp was  68.45
City found , max temp was  71.01
City found , max temp was  14
City found , max temp was  64.4
City found , ma

City found , max temp was  80.6
City found , max temp was  73.63
City found , max temp was  75.2
City found , max temp was  6.8
City found , max temp was  3.99
City found , max temp was  51.98
City found , max temp was  69.49
City found , max temp was  41.25
City found , max temp was  73.33
City found , max temp was  76.19
City not found. Skipping...
City found , max temp was  32.38
City found , max temp was  69.93
City found , max temp was  77
City found , max temp was  64.53
City found , max temp was  60.8
City found , max temp was  44.6
City not found. Skipping...
City found , max temp was  53.83
City found , max temp was  24.01
City found , max temp was  43.16
City not found. Skipping...
City found , max temp was  64.36
City found , max temp was  77.59
City found , max temp was  78.8
City found , max temp was  16.48
City found , max temp was  74.14
City found , max temp was  32
City found , max temp was  63
City found , max temp was  5.29
City found , max temp was  48.99
City found

In [74]:
len(cities)

597

In [75]:
len(city_data)

534

# "ETL" Steps : Extract, Transform, Load
## Pre-processing your Data

 - Latitudes and longitudes; randomly generated; 1500
 - We used an citipy API (Python library / input-output) to take input: lat, lon, output: city names; 597 (less than 1500, because of "uniqueness")
 - We used a openweathermap API (web API / request-response) to take input: city name, output: weather info; 534 (less than 597; because the API didn't know about those cities)

In [76]:
# Convert the array of dictionaries to a Pandas DataFrame.
city_data_df= pd.DataFrame(city_data)
city_data_df.head(10)

City    Lat     Lng  Max Temp  Humidity  Cloudiness  Wind Speed  \
0     Henties Bay -22.12   14.28     56.66        91          26        1.19   
1          Bethel  60.79 -161.76     37.40        90          90        9.17   
2  Saint-Francois  46.42    3.91     46.40        81          98        6.93   
3         Ushuaia -54.81  -68.31     33.80        98          90       27.51   
4   Mar Del Plata -46.43  -67.52     49.62        31           0       33.98   
5    Kruisfontein -34.00   24.73     56.61        80          97       24.83   
6     Bronnoysund  65.46   12.20     32.00        64         100       11.41   
7           Yumen  40.29   97.04     44.55        41           2       23.35   
8          Albany  42.65  -73.75     39.00        55          90        3.36   
9          Dikson  73.51   80.55    -11.64        93           1       11.81   

  Country                 Date  
0      NA  2019-11-07 02:59:25  
1      US  2019-11-07 02:59:26  
2      FR  2019-11-07 02:58:41  
3      AR  2019-11-07 02:59:26  
4      AR  2019-11-07 02:58:41  
5      ZA  2019-11-07 02:58:41  
6      NO  2019-11-07 02:59:27  
7      CN  2019-11-07 02:59:27  
8      US  2019-11-07 02:59:28  
9      RU  2019-11-07 02:59:28

In [77]:
# Create the output file (CSV).
output_data_file = "cities.csv"
# Export the City_Data into a CSV.
city_data_df.to_csv(output_data_file, index_label="City_ID")


In [78]:
#city_data_df
len(city_data_df)
city_data_df.count()
city_data_df

City    Lat     Lng  Max Temp  Humidity  Cloudiness  \
0       Henties Bay -22.12   14.28     56.66        91          26   
1            Bethel  60.79 -161.76     37.40        90          90   
2    Saint-Francois  46.42    3.91     46.40        81          98   
3           Ushuaia -54.81  -68.31     33.80        98          90   
4     Mar Del Plata -46.43  -67.52     49.62        31           0   
..              ...    ...     ...       ...       ...         ...   
529       Vredendal -31.68   18.49     54.90        79          10   
530         Kushima  31.50  131.29     69.80        52          75   
531        Eirunepe  -6.66  -69.87     72.54        98          17   
532       Tocopilla -22.09  -70.20     59.07        36          33   
533     Verkhoyansk  67.55  133.39     -5.68        95         100   

     Wind Speed Country                 Date  
0          1.19      NA  2019-11-07 02:59:25  
1          9.17      US  2019-11-07 02:59:26  
2          6.93      FR  2019-11-07 02:58:41  
3         27.51      AR  2019-11-07 02:59:26  
4         33.98      AR  2019-11-07 02:58:41  
..          ...     ...                  ...  
529        6.96      ZA  2019-11-07 03:01:44  
530        6.93      JP  2019-11-07 03:01:44  
531        1.45      BR  2019-11-07 03:01:45  
532        8.10      CL  2019-11-07 03:01:45  
533        2.21      RU  2019-11-07 03:01:45  

[534 rows x 9 columns]